pkg/log/zap/flags.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zap
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/spf13/pflag"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var (
+	zapFlagSet *pflag.FlagSet
+
+	development bool
+	encoderVal  encoderValue
+	levelVal    levelValue
+	sampleVal   sampleValue
+)
+
+func init() {
+	zapFlagSet = pflag.NewFlagSet("zap", pflag.ExitOnError)
+	zapFlagSet.BoolVar(&development, "zap-devel", false, "Enable zap development mode (changes defaults to console encoder, debug log level, and disables sampling)")
+	zapFlagSet.Var(&encoderVal, "zap-encoder", "Zap log encoding ('json' or 'console')")
+	zapFlagSet.Var(&levelVal, "zap-level", "Zap log level (one of 'debug', 'info', 'warn', 'error', 'dpanic', 'panic', 'fatal')")
```

Comment:
I think you're correct. I'll verify and fix.

Reply:
Just checked and found a few things:
1. `Error()` is part of the `logr.Logger` interface, so at a minimum we would need `debug`, `info`, and `error`.
2. It's possible to get handles to each of the log levels using the `logr.Logger` interface :
    ```go
    log.V(2).Info("DEBUG Level 2")
    log.V(1).Info("DEBUG Level 1")
    log.V(0).Info("INFO Level (with specific verbosity number)")
    log.Info("INFO Level (without specific verbosity number)")
    log.V(-1).Info("WARN Level")
    log.V(-2).Info("", "error", "ERROR Level (with specific verbosity number)")
    log.Error(errors.New("ERROR Level (without specific verbosity number)"), "")
    log.V(-3).Info("DPANIC Level")
    log.V(-4).Info("PANIC Level")
    log.V(-5).Info("FATAL Level")
    ```
3. It is not currently possible to set a log level to get `log.V(2)` (or higher DEBUG levels) to show up. Only `log.V(1)` will output debug logs. If we allow integers for the `--zap-level` flag, we could probably make higher DEBUG levels work.

Reply:
Makes sense. Having integers for 0,1,2,3... is fine.
Logr discourages using anything other than info, error and integer debug levels so ideally users just stick to the integers.

But if for some reason users want to set the minimum level to `warn` that skips all info and debug logs, well I guess we'll allow that.

Reply:
For the record, I'm not opposed to slimming the accepted values down to the following to be more in line with logr:
* `error`
* `info`/`0`
* `debug`/`1`
* `2`
* `3`
* etc...

At that point we could probably name the flag `--log-level`, since it would be applicable to any logr implementation.

Reply:
@hasbro17 now we're back full circle. Hitting this panic trying to set the log level to higher debug levels: https://github.com/kubernetes-sigs/controller-runtime/issues/167#issuecomment-426963327

[Zap has a well-defined set of levels between `debug` and `fatal`,](https://github.com/uber-go/zap/blob/82e3b47c1fe49387143b1a051a9ab6f34a08678d/zapcore/level.go#L34-L56) so it makes sense that trying to use a log level outside of that range would fail.

I guess this means there's no need to support integers for the flag since it is zap-specific? Any opinions on using all of them or just the ones relevant to `logr`? I think I lean toward only the `logr` ones (`debug`, `info`, and `error`).

I'll find some place to document that `log.V(2)` (or higher) is not supported by the zap logger.

Reply:
@joelanford I forgot about that bug. I was sure that zap supports custom debug levels like `v(-2)` and that bug was actually in zapr. But I'll need to take a closer look.

For now I agree that we can keep this simple and just provide `debug`, `info`, and `error` as configurable levels and be in line with logr conventions. And we can document `log.V(2)` as not being supported.